src/main/java/org/broadinstitute/hellbender/engine/TwoPassReadWalker.java

```diff
@@ -15,6 +15,10 @@
  * optional contextual information from a reference and/or sets of variants/Features. The TwoPassReadsWalker iterates
  * through the reads twice, allowing for different processing to be performed on the reads each time.
  *
+ * WARNING: This traversal should only be used in the rare case that complex state must be maintained between two passes
+ * over the reads. In most cases it's preferable to either split the two passes into separate walkers or to restructure
+ * the algorithm so that it's not necessary to make multiple passes.
+ *
  * If multiple sources of reads are specified, they are merged together into a single sorted stream of reads.
  *
  * TwoPassReadWalker authors must implement the firstPassApply() and secondPassApply() methods to process each read,
```

Comment:
Thinking about it, there should probably be a link in this description that references the ReadWalker class and makes note that this is analogous.